### ISSUE TITLE
UPX v4.1.0 Task List

### ISSUE THREAD

**Issue by markus-oberhumer** [upx/upx#684]
Experiment: using a GitHub issue+tasklists; this will be updated frequently; feel free to add comments; comments that are done will be hidden/collapsed; no actual discussions here, use existing or new issues instead!

### UPX v4.1.0 Milestone
https://github.com/upx/upx/milestone/11

### 4 PT_LOAD segments
- [x] are we fully compatible unpacking all binaries packed by previous upx versions?
 TESTED: yes, with a single exception: `upx-3.94-armeb_linux/upx`; upx-4.0.2 can unpack this file, but current devel cannot; FIXED
- [x] can old upx versions unpack the new 4 PT_LOAD files or give meaningful error messages? jreiser TESTED: yes

### Android
- [x] https://github.com/upx/upx/issues/220
- [x] https://github.com/upx/upx/issues/609
- [x] https://github.com/upx/upx/issues/671
- [x] https://github.com/upx/upx/issues/673
- [x] https://github.com/upx/upx/issues/680
- [x] https://github.com/upx/upx/issues/693

### Unsorted
- [x] update file `NEWS`
- [x]  spell check
- [x] `make run-testsuite`: there are some warnings during the `t010_decompressed` step; INTENDED
- [x] `make run-testsuite` t010_decompressed currently does not work for big endian builds like mips or armeb (when using qemu); FIXED

**Comment by jreiser** [comment]
@markus-oberhumer   re: `make run-testsuite t010_decompressed` currently does not work for big endian builds like mips or armeb (when using qemu)   

Please be more specific than  "does not work"; please give a reproducible test case. Both mips and armeb work for me under patched qemu, where `upx.fix` is the result of applying the upx from the commit below to the de-compressed upx in the 4.0.2 release directories upx-4.0.2-mips_linux and upx-4.0.2-armeb_linux:

commit 067895fa5f35840e2406abcdd9a2e25bf0871b37 (origin/devel)
Date:   Sun Jul 16 16:39:52 2023 -0700

```
$ qemu/build/qemu-mips upx.fix -d -f -o foo upx.fix
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2023
UPX 4.0.2       Markus Oberhumer, Laszlo Molnar & John Reiser   Jan 30th 2023

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
   2493528 <-    745876   29.91%   linux/mips    foo

Unpacked 1 file.
$ cd ../upx-4.0.2-armeb_linux
$ qemu/build/qemu-armeb upx.fix -d -f -o foo upx.fix
                       Ultimate Packer for eXecutables
                          Copyright (C) 1996 - 2023
UPX 4.0.2       Markus Oberhumer, Laszlo Molnar & John Reiser   Jan 30th 2023

        File size         Ratio      Format      Name
   --------------------   ------   -----------   -----------
   2005288 <-    641124   31.97%   linux/armeb   foo

Unpacked 1 file.
$ 
```

**Comment by markus-oberhumer** [comment]
@jreiser Don't be such a slacker, just run the upx testsuite under qemu-mips !!

(you can use the compiler from https://github.com/upx/upx-stubtools/releases/download/v20160918/upx-linux-musl-gcc-9.2.0-toolchains-20200226.tar.xz or try precompiled UPX binaries from https://github.com/upx/upx/actions/workflows/weekly-ci-zigcc.yml)

**Comment by markus-oberhumer** [comment]
@jreiser Please indicate your preference if we should move some tasks to v4.2.0 so that we can release v4.1.0 soon

**Comment by jreiser** [comment]
Move all MacOS out of v4.1.0 now, except for releasing builds that run on MacOS.

I think I'm somewhat close on Android, but users have not responded to the latest build of more than one week ago, and getting an emulator to install+run still is problematic, and rooting my Android tablets (so that I can run tests on real Android) still has not succeeded.  Allow another 10 days; if Android still balks, then move Android out of v4.1.0.

`just run the upx testsuite under qemu-mips`  Does this mean use `qemu-mips upx-4.0.2-mips_linux/upx` instead of some version of upx-x86_64 when running `make run-testsuite`?

**Comment by markus-oberhumer** [comment]
No, releasing official macOS binaries for the first time while UPX does not work on macOS 13 is just bad marketing.

**Comment by jreiser** [comment]
Very well; but making a upx-compressed executable run on MacOS is going to take some months of doing nothing else but figuring out the actual minimal requirements for any executable to run on MacOS.  I've even thought about making a separate dynamic loader `/opt/bin/upx-dyld`  as an intermediate step for development, because dyld can be built once "by hand" to meet all the requirements, instead of having to generate them all when compressing.

**Comment by markus-oberhumer** [comment]
> `just run the upx testsuite under qemu-mips` Does this mean use `qemu-mips upx-4.0.2-mips_linux/upx` instead of some version of upx-x86_64 when running `make run-testsuite`?

Exactly. This works for me:

```sh
cd some-tmp-dir

# IMPORTANT: make sure that paths are absolute in $upx_exe
export upx_exe="qemu-mips $HOME/code/github/upx/upx4/build/official/mips-linux.musl-gcc-9.0/release/upx"

export upx_testsuite_SRCDIR=$HOME/code/github/upx/upx-testsuite.git

# info: this fails for me with big-endian qemu-mips
bash ~/code/github/upx/upx4/misc/testsuite/upx_testsuite_1.sh

ls -la ./tmp-upx-testsuite
```

**Comment by markus-oberhumer** [comment]
@jreiser it seems your `upx-3.94-armeb_linux` fix has also fixed the mips big-endian `make run-testsuite` `t010_decompressed` issue. Nice, planned or side effect?

**Comment by jreiser** [comment]
@markus-oberhumer  Both machines are big endian, both could suffer ["enjoy"] the episodes of "b_info corrupted ... recovery at".  So, not really planned.  Will check tomorrow due to instability in qemu.

I found it hard to follow the flow in `misc/testsuite/upx_testsuite_1.sh`.  It's very efficient, but it is hard for me to pin down two explicit commands that use different approaches to compute the same goal, but produce different actual results. In particular, I'm wondering how `upx_testsuite_1.sh` avoids the problem that different versions of upx **always** will produce different results if the stubs changed between the versions.  If development commits a change to `upx_testsuite_1-expected_sha256sums.sh` then that difference propagates to any compression.

**Comment by markus-oberhumer** [comment]
I'm not fully sure what you mean, but upx_testsuite_1.sh gets the expected checksums with this line:

```sh
source "$argv0dir/upx_testsuite_1-expected_sha256sums.sh" || exit 1
```

**Comment by jreiser** [comment]
The checksums in `upx_testsuite_1-expected_sha256sums.sh` are valid only for the version of upx that created them.  Any other version of upx almost certainly will produce different checksums, because of changes to `src/stub/src/...` and the resulting "sync checksums" commit to `misc/testsuite/upx_testsuite_1-expected_sha256sums.sh`.  This is why I'm so concerned with producing two **explicit** command lines that supposedly should produce the same output, but actually do not.  Only then can debugging begin.

**Comment by markus-oberhumer** [comment]
I fear I still don't get you -  the simple upx_testsuite_1 script just checks that the current version creates results that do match the expected checksums. This covers general functionality and some basic BE vs LE and 32 vs 64-bit checks.

**Comment by markus-oberhumer** [comment]
@jreiser anything you want to add to `NEWS` ?

**Comment by jreiser** [comment]
@markus-oberhumer  Nothing from me to add to NEWS.  Getting the Android bugs fixed will be enough.

**Comment by markus-oberhumer** [comment]
@jreiser devel4 review

```
$ git fetch --all --prune
$ git diff 56ddfac3..origin/devel4-mfx-review
```

**Comment by jreiser** [comment]
@markus-oberhumer  Fixed. and pushed.  Thanks!

**Comment by jreiser** [comment]
`can old upx versions unpack the new 4 PT_LOAD files`  I tried the releases listed in https://github.com/upx/upx/releases , using a compressed version of /bin/date from Fedora 36-x86_64, where the original has 4 PT_LOAD segments.
Versions 4.0.2, 4.0.1, 4.0.0 all unpack correctly (unpacked file matches original file).
Versions 3.96, 3.95, 3.94, 3.93, 3.92 all complain `CantUnpackException: need a newer version of UPX`.
Version 3.91 gets SIGSEGV.
Version 3.00 has no upx in Artifacts, and I decided not to build it.

So, I claim success on this task.

**Comment by markus-oberhumer** [comment]
@jreiser new devel4 branch has been pushed, don't forget `git submodule update` after pull

**Comment by markus-oberhumer** [comment]
@jreiser please have a look at the current clang-analyzer log https://github.com/upx/upx/actions/runs/5664916977/job/15348935773 (you can ignore the deadcode.DeadStores warnings, but the core.CallAndMessage needs to be checked)

EDIT: you can run the clang-analyzer locally by using `make build/extra/scan-build/release` (see top-level Makefile for more info), but results may differ from the Alpine Linux version

**Comment by jreiser** [comment]
@markus-oberhumer  Commit 400cc293a0d75389806f16064f092ccc0f955626 addresses CallAndMessage, plus many of DeadStore.

I tried to run scan-build locally, but scan-build could not be found:
```
$ cd $(top_level); make build/extra/scan-build/release
scan-build cmake -S . -B build/extra/scan-build/release  -DCMAKE_BUILD_TYPE=Release
make: scan-build: No such file or directory
make: *** [Makefile:128: build/extra/scan-build/release] Error 127
```
So I did
```
$ mkdir -p build/extra/scan-build/release; cd !$
$ cmake ../../..
$ cmake --build .
```
which did a regular build.  Still no scan-build.

Oh, scan-build is in package clang-analyzer.  It should be called clang-scan-build.  Typical poor user interface from clang.

**Comment by markus-oberhumer** [comment]
@jreiser I'm currently working on some more analyzer improvements, in the meantime please install the binary clang-tidy (might be in package `clang-tools-extra`)

**Comment by markus-oberhumer** [comment]
@jreiser another new devel4 branch has been pushed, don't forget git submodule update after pull once again

Please have a look at the 4 warnings from clang-tidy
https://github.com/upx/upx/actions/runs/5673844096/job/15376190885

And you also should be able to run clang-tidy locally: `make build/analyze/clang-tidy/release`

**Comment by markus-oberhumer** [comment]
Looks like all tasks are closed, nice. Is there anything you want to work on before releasing v4.1.0 ?

**Comment by markus-oberhumer** [comment]
In any case I will wait until the milestone issues are closed.

**Comment by markus-oberhumer** [comment]
Finally I also think this GitHub issue+tasklist experiment was a success.

**Comment by markus-oberhumer** [comment]
@jreiser Please have a look at Spell check
https://github.com/upx/upx/actions/runs/5704946488/job/15459065252

**Comment by jreiser** [comment]
https://github.com/upx/upx/commit/222aac44a0ac34a65efe7aedb59864d57c107d53     spell check typos

**Comment by markus-oberhumer** [comment]
What is the status of #671 ? Should I move this to v4.2.0 ?

**Comment by jreiser** [comment]
The status of #671 is waiting for hardware to try the fixes (or discover that duckdb use self-inspection that is "illegal".)  My two Samsung tablet are in limbo because "rooting" failed and recovery is painful.  I have another tablet being delivered today.  Supposedly duckdb under termux does not require rooting, so I should be able to test with ordinary "developer access" in user mode.  If that fails, or takes too long, then this issue should move to v4.2.0.

**Comment by jreiser** [comment]
Issue #671 has been closed as `invalid` because `proot` did not change the Linux filename `libdl.so.2` to the corresponding Android name `libdl.so` (with no suffix ".2").  So all Issues of Milestone 4.1.0 have been closed.

**Comment by markus-oberhumer** [comment]
Thanks John! Assuming no commits from your side are pending I'll make the v4.1.0 release in the next days.

**Comment by jreiser** [comment]
I'm all caught up (notice the "HEAD ->" branches):
```
$ git branch
* devel4
$ git log -n 1
commit 222aac44a0ac34a65efe7aedb59864d57c107d53 (HEAD -> devel4, origin/devel4, origin/HEAD)
Date:   Mon Jul 31 09:05:32 2023 -0700

    spell check typos
```

**Comment by markus-oberhumer** [comment]
#693 emerged, and I won't have time on this weekend anyway

**Comment by jreiser** [comment]
https://github.com/upx/upx/issues/693 fixed.
